File: src/lib/matcher.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import type { FileInfo } from './types';

export type NasIndex = Map<string, string[]>;

// System / recycle-bin folders that should never be scanned or suggested.
// Synology's @eaDir thumbnail dirs in particular shadow original filenames
// and make SMB walks several times slower.
const IGNORED_FOLDERS = new Set(
  ['$RECYCLE.BIN', 'System Volume Information', '#recycle', '@eaDir', '.Trashes', 'found.000']
    .map((n) => n.toLowerCase()),
);

const STAT_CONCURRENCY = 16;

function makeKey(name: string, size: number): string {
  return `${name}|${size}`;
}

export async function indexNas(
  nasPath: string,
  onProgress?: (count: number, folder: string) => void,
  targetKeys?: Set<string>,
): Promise<NasIndex> {
  const index: NasIndex = new Map();
  const remaining = targetKeys ? new Set(targetKeys) : null;
  await walkNas(nasPath, index, remaining, onProgress);
  return index;
}

/** Returns true when all target keys have been found (caller can stop). */
async function walkNas(
  dir: string,
  index: NasIndex,
  remaining: Set<string> | null,
  onProgress?: (count: number, folder: string) => void,
): Promise<boolean> {
  let entries: fs.Dirent[];
  try {
    entries = await fs.promises.readdir(dir, { withFileTypes: true });
  } catch {
    return false;
  }

  if (onProgress) onProgress(index.size, path.basename(dir));

  const subdirs: string[] = [];
  const fileNames: string[] = [];

  for (const entry of entries) {
    if (entry.name.startsWith('.') || IGNORED_FOLDERS.has(entry.name.toLowerCase())) continue;
    if (entry.isDirectory()) {
      subdirs.push(path.join(dir, entry.name));
    } else if (entry.isFile()) {
      fileNames.push(entry.name);
    }
  }

  // Stat files with bounded concurrency — sequential stats over SMB are
  // one network round trip each and dominate scan time.
  let next = 0;
  const statWorker = async () => {
    while (next < fileNames.length) {
      const name = fileNames[next++];
      try {
        const stat = await fs.promises.stat(path.join(dir, name));
        const key = makeKey(name, stat.size);
        const existing = index.get(key);
        if (existing) existing.push(dir);
        else index.set(key, [dir]);
        remaining?.delete(key);
      } catch {
        // skip
      }
    }
  };
  await Promise.all(
    Array.from({ length: Math.min(STAT_CONCURRENCY, fileNames.length) }, statWorker),
  );

  if (remaining && remaining.size === 0) return true;

  // Sort descending so newest folders are scanned first
  subdirs.sort().reverse();
  for (const sub of subdirs) {
    if (await walkNas(sub, index, remaining, onProgress)) return true;
  }
  return false;
}

export interface SourceIndex {
  name: string;
  index: NasIndex;
}

export interface SuggestedFolder {
  folder: string;
  count: number;
  source: string;
  newestMtime: number; // newest mtime among the matched SD files (recency signal)
}

// Walk up to the nearest ancestor whose name starts with YYYY.MM.DD or YYYY-MM-DD
function toDateLevelFolder(folderPath: string): string {
  const parts = folderPath.split(path.sep);
  for (let i = parts.length - 1; i >= 0; i--) {
    if (/^\d{4}[.\-]\d{2}[.\-]\d{2}/.test(parts[i])) {
      return parts.slice(0, i + 1).join(path.sep);
    }
  }
  return folderPath;
}

export function checkBackedUp(
  sdFiles: FileInfo[],
  sourceIndexes: SourceIndex[],
): { backedUp: FileInfo[]; missing: FileInfo[]; suggestedFolders: SuggestedFolder[] } {
  const merged = mergeIndexes(...sourceIndexes.map((s) => s.index));
  const backedUp: FileInfo[] = [];
  const missing: FileInfo[] = [];
  // Track folder → { count, source, newestMtime }
  const folderInfo = new Map<string, { count: number; source: string; newestMtime: number }>();

  // Build folder→source lookup (also index normalized date-level paths)
  const folderToSource = new Map<string, string>();
  for (const si of sourceIndexes) {
    for (const [, paths] of si.index) {
      for (const p of paths) {
        if (!folderToSource.has(p)) folderToSource.set(p, si.name);
        const top = toDateLevelFolder(p);
        if (!folderToSource.has(top)) folderToSource.set(top, si.name);
      }
    }
  }

  for (const f of sdFiles) {
    const key = makeKey(f.name, f.size);
    const paths = merged.get(key);
    if (paths) {
      backedUp.push(f);
      for (const p of paths) {
        const top = toDateLevelFolder(p);
        const existing = folderInfo.get(top);
        if (existing) {
          existing.count++;
          existing.newestMtime = Math.max(existing.newestMtime, f.mtime ?? 0);
        } else {
          folderInfo.set(top, {
            count: 1,
            source: folderToSource.get(p) ?? folderToSource.get(top) ?? 'Unknown',
            newestMtime: f.mtime ?? 0,
          });
        }
      }
    } else {
      missing.push(f);
    }
  }

  const suggestedFolders = [...folderInfo.entries()]
    .map(([folder, { count, source, newestMtime }]) => ({ folder, count, source, newestMtime }))
    .sort((a, b) => b.count - a.count)
    .slice(0, 10);

  return { backedUp, missing, suggestedFolders };
}

export function mergeIndexes(...indexes: NasIndex[]): NasIndex {
  const merged: NasIndex = new Map();
  for (const idx of indexes) {
    for (const [key, paths] of idx) {
      const existing = merged.get(key);
      if (existing) existing.push(...paths);
      else merged.set(key, [...paths]);
    }
  }
  return merged;
}

export function listExistingFolders(nasPath: string): string[] {
  try {
    const entries = fs.readdirSync(nasPath, { withFileTypes: true });
    return entries
      .filter((e) => e.isDirectory() && !e.name.startsWith('.') && !IGNORED_FOLDERS.has(e.name.toLowerCase()))
      .map((e) => e.name)
      .sort();
  } catch {
    return [];
  }
}

File: macos/CopierCore/Sources/CopierCore/Matcher.swift
import Foundation

/// A file-name+size index of one location, mapping each key to the folders that hold it.
public struct LocationIndex: Sendable, Hashable {
    public private(set) var entries: [FileKey: [URL]]

    public init(entries: [FileKey: [URL]] = [:]) {
        self.entries = entries
    }

    public var count: Int { entries.count }

    public mutating func add(_ key: FileKey, folder: URL) {
        entries[key, default: []].append(folder)
    }

    public func folders(for key: FileKey) -> [URL]? { entries[key] }

    /// Merge several indexes into one.
    public static func merged(_ indexes: [LocationIndex]) -> LocationIndex {
        var merged = LocationIndex()
        for index in indexes {
            for (key, folders) in index.entries {
                merged.entries[key, default: []].append(contentsOf: folders)
            }
        }
        return merged
    }
}

/// An index together with the name of the source it came from.
public struct SourceIndex: Sendable {
    public var name: String
    public var index: LocationIndex

    public init(name: String, index: LocationIndex) {
        self.name = name
        self.index = index
    }
}

/// A destination folder that already holds files from this card.
public struct SuggestedFolder: Sendable, Hashable {
    public var folder: URL
    public var count: Int
    public var source: String
    /// Newest modification date among the matched card files (recency signal).
    public var newestModification: Date?

    public init(folder: URL, count: Int, source: String, newestModification: Date?) {
        self.folder = folder
        self.count = count
        self.source = source
        self.newestModification = newestModification
    }
}

/// Result of comparing the card against all check sources.
public struct MatchResult: Sendable {
    /// Files already present in at least one location (name + exact size).
    public var backedUp: [MediaFile]
    /// Files not found anywhere.
    public var missing: [MediaFile]
    /// Up to ten destination folders that already hold files from this card, most matches first.
    public var suggestedFolders: [SuggestedFolder]

    public init(backedUp: [MediaFile], missing: [MediaFile], suggestedFolders: [SuggestedFolder]) {
        self.backedUp = backedUp
        self.missing = missing
        self.suggestedFolders = suggestedFolders
    }
}

/// Duplicate detection: a file counts as backed up when its name and exact byte
/// size are found in any indexed location.
public enum Matcher {
    /// System / recycle-bin folders that are never scanned or suggested.
    /// Synology's `@eaDir` thumbnails in particular shadow original names and slow SMB walks down.
    public static let ignoredFolders: Set<String> = [
        "$recycle.bin", "system volume information", "#recycle", "@eadir", ".trashes", "found.000",
    ]

    /// Number of parallel `stat` calls per folder — sequential stats over SMB are one
    /// network round trip each and dominate scan time.
    public static let statConcurrency = 16

    /// `true` when a directory name must be skipped during indexing.
    public static func isIgnored(directoryName name: String) -> Bool {
        name.hasPrefix(".") || ignoredFolders.contains(name.lowercased())
    }

    /// Index a local (or SMB-mounted) path.
    ///
    /// Newest folders are visited first, and the walk stops early once every key in
    /// `targetKeys` has been found.
    public static func indexLocalPath(
        _ root: URL,
        targetKeys: Set<FileKey>? = nil,
        progress: (@Sendable (ScanProgress) -> Void)? = nil
    ) async -> LocationIndex {
        var index = LocationIndex()
        var remaining = targetKeys
        await walk(root, index: &index, remaining: &remaining, progress: progress)
        return index
    }

    @discardableResult
    private static func walk(
        _ directory: URL,
        index: inout LocationIndex,
        remaining: inout Set<FileKey>?,
        progress: (@Sendable (ScanProgress) -> Void)?
    ) async -> Bool {
        if Task.isCancelled { return true }

        let keys: [URLResourceKey] = [.isDirectoryKey, .isRegularFileKey, .fileSizeKey]
        let entries = (try? FileManager.default.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: keys,
            options: []
        )) ?? []

        progress?(ScanProgress(count: index.count, folder: directory.lastPathComponent))

        var subdirectories: [URL] = []
        var fileURLs: [URL] = []
        for entry in entries {
            let name = entry.lastPathComponent
            if isIgnored(directoryName: name) { continue }
            guard let values = try? entry.resourceValues(forKeys: Set(keys)) else { continue }
            if values.isDirectory == true {
                subdirectories.append(entry)
            } else if values.isRegularFile == true {
                fileURLs.append(entry)
            }
        }

        let sized = await sizes(of: fileURLs)
        for (name, size) in sized {
            let key = FileKey(name: name, size: size)
            index.add(key, folder: directory)
            remaining?.remove(key)
        }

        if let remaining, remaining.isEmpty { return true }

        // Descending so the newest (date-named) folders are scanned first.
        for sub in subdirectories.sorted(by: { $0.lastPathComponent > $1.lastPathComponent }) {
            if await walk(sub, index: &index, remaining: &remaining, progress: progress) { return true }
        }
        return false
    }

    /// Stat a batch of files with bounded concurrency.
    private static func sizes(of urls: [URL]) async -> [(String, Int64)] {
        guard !urls.isEmpty else { return [] }
        let limit = min(statConcurrency, urls.count)
        return await withTaskGroup(of: (String, Int64)?.self) { group in
            var next = 0
            func submit() {
                let url = urls[next]
                next += 1
                group.addTask {
                    guard let values = try? url.resourceValues(forKeys: [.fileSizeKey]),
                          let size = values.fileSize
                    else { return nil }
                    return (url.lastPathComponent, Int64(size))
                }
            }
            while next < limit { submit() }
            var result: [(String, Int64)] = []
            while let finished = await group.next() {
                if let finished { result.append(finished) }
                if next < urls.count { submit() }
            }
            return result
        }
    }

    /// Split the card files into backed-up and missing, and suggest destination folders.
    public static func checkBackedUp(
        files: [MediaFile],
        sources: [SourceIndex]
    ) -> MatchResult {
        let merged = LocationIndex.merged(sources.map(\.index))

        var folderToSource: [URL: String] = [:]
        for source in sources {
            for (_, folders) in source.index.entries {
                for folder in folders {
                    if folderToSource[folder] == nil { folderToSource[folder] = source.name }
                    let top = dateLevelFolder(folder)
                    if folderToSource[top] == nil { folderToSource[top] = source.name }
                }
            }
        }

        var backedUp: [MediaFile] = []
        var missing: [MediaFile] = []
        var info: [URL: (count: Int, source: String, newest: Date?)] = [:]

        for file in files {
            guard let folders = merged.folders(for: FileKey(file)) else {
                missing.append(file)
                continue
            }
            backedUp.append(file)
            for folder in folders {
                let top = dateLevelFolder(folder)
                if var existing = info[top] {
                    existing.count += 1
                    existing.newest = newer(existing.newest, file.modificationDate)
                    info[top] = existing
                } else {
                    info[top] = (
                        1,
                        folderToSource[folder] ?? folderToSource[top] ?? "Unknown",
                        file.modificationDate
                    )
                }
            }
        }

        var suggested: [SuggestedFolder] = []
        suggested.reserveCapacity(info.count)
        for (folder, value) in info {
            suggested.append(
                SuggestedFolder(
                    folder: folder,
                    count: value.count,
                    source: value.source,
                    newestModification: value.newest
                )
            )
        }
        suggested.sort { (lhs: SuggestedFolder, rhs: SuggestedFolder) -> Bool in
            if lhs.count != rhs.count { return lhs.count > rhs.count }
            return lhs.folder.path < rhs.folder.path
        }

        return MatchResult(backedUp: backedUp, missing: missing, suggestedFolders: Array(suggested.prefix(10)))
    }

    private static func newer(_ lhs: Date?, _ rhs: Date?) -> Date? {
        switch (lhs, rhs) {
        case let (l?, r?): return max(l, r)
        case let (l?, nil): return l
        case let (nil, r?): return r
        default: return nil
        }
    }

    /// Walk up to the nearest ancestor whose name starts with `YYYY.MM.DD` or `YYYY-MM-DD`.
    public static func dateLevelFolder(_ folder: URL) -> URL {
        let components = folder.standardizedFileURL.pathComponents
        for index in stride(from: components.count - 1, through: 0, by: -1) {
            if startsWithDate(components[index]) {
                var result = URL(fileURLWithPath: "/")
                for component in components[1...index] {
                    result.appendPathComponent(component)
                }
                return result
            }
        }
        return folder
    }

    private static func startsWithDate(_ name: String) -> Bool {
        let chars = Array(name)
        guard chars.count >= 10 else { return false }
        func digits(_ range: Range<Int>) -> Bool { range.allSatisfy { chars[$0].isNumber } }
        let separators: Set<Character> = [".", "-"]
        return digits(0..<4) && separators.contains(chars[4]) && digits(5..<7)
            && separators.contains(chars[7]) && digits(8..<10)
    }

    /// Folder names directly under `root`, ignoring dot-folders and system folders.
    public static func listExistingFolders(at root: URL) -> [String] {
        let entries = (try? FileManager.default.contentsOfDirectory(
            at: root,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: []
        )) ?? []
        return entries
            .filter { (try? $0.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) == true }
            .map(\.lastPathComponent)
            .filter { !isIgnored(directoryName: $0) }
            .sorted()
    }

}
